# Homework 3  - Tryout pandas and numpy

Copy this notebook. Rename it as: YOURNAME-Homework-panda-numpy-Sept-2017

with your name replacing yourname and  replaced with the date you submit or to the last part.

Do the homeworks problems below.

Upload your completed jupyter notebook to your github site and send me the url via the elearning site as your homework submission. Please note that you need to do your own 3.4, and 3.5 if you share data and analysis from others then you cannot get more than a 3.  

## Homework 3.1

### 3.1.a
Load the data from: http://opendata.dc.gov/datasets that I have include in this github   
into a dataframe.  ( The file has been is available in directory ./data/ccp_current_csv.csv )

In [1]:
import pandas as pd
import numpy as np
ccp = pd.read_csv('data/ccp_current_csv.csv')
ccp.head()

STREET QUADRANT  FROMINTERSECTION              TOINTERSECTION  \
0           13th St       NW        Kennedy St               Longfellow St   
1              S St       NW  North Capitol St                      3rd St   
2  Rhode Island Ave       NE  Brentwood Rd (W)            Brentwood Rd (M)   
3           Mill Rd       NW           27th St                  Cul De Sac   
4  Rhode Island Ave       NE  Brentwood Rd (M)  14th St  Evarts St  Montan   

   NOOFBLOCKS                 WORKDESCRIPTION  YEARBUDGETED  PERCENTCOMPLETED  \
0           1  Pavement Resurface/restoration          2004               100   
1           3                   Regular Cover          1999               100   
2           1  Pavement Resurface/restoration          2004               100   
3           1  Pavement Resurface/restoration          2004               100   
4           2  Pavement Resurface/restoration          2005               100   

      STATUS  MILES      ...             YCOORD   LATITUDE  LONGITUDE  \
0  Completed   0.07      ...       143205.11981  38.956747 -77.029694   
1  Completed   0.33      ...       138468.90256  38.914085 -77.012411   
2  Completed   0.09      ...       139574.10659  38.924041 -76.987221   
3  Completed   0.07      ...       138132.90198  38.911046 -77.054077   
4  Completed   0.16      ...       139657.25318  38.924790 -76.985068   

           PROJECTNOTE                                        PROJECTNAME  \
0                  NaN                                             FedAid   
1                  NaN                                                NaN   
2                  NaN           Fedral Aid Citywide Pavement Restoration   
3  POKA-2003-B-0048-JJ  Citywide Pavement Restoration Contract (Local ...   
4                  NaN                                                NaN   

          PROJECTMGR  PROJECTID        TEAMREP                 EMAIL  \
0    Paul Stephenson       5834        O. Hill                   NaN   
1                NaN       2467            NaN                   NaN   
2  Mohamed Abdullahi       5486   Said Cherifi   said.cherifi@dc.gov   
3    Maduabuchi Udeh       5796  Michael Jelen  Michael.Jelen@dc.gov   
4                NaN       8237            NaN                   NaN   

   CONTACTPHONE  
0  202-671-4591  
1           NaN  
2  202-671-4611  
3  202-671-4542  
4           NaN  

[5 rows x 27 columns]

### 3.1.a  what is its shape and what does that mean?

In [2]:
ccp.shape

(465, 27)

The shape is the number of columns and number of rows. For this dataset the shape is 465, 27. So there are 465 rows and 27 columns.

## Homework 3.2
 What are the number of rows in each 'QUADRANT' ?

In [3]:
pd.value_counts(ccp.QUADRANT)

NW    195
NE    163
SE     84
BN     12
SW     11
Name: QUADRANT, dtype: int64

The number of rows in each quadrant is:
NW: 195, NE: 163, SE: 84, BN: 12, SW: 11

## Homework 3.3 - Array math demonstration 
For two arrarys  

a= [1,2,3,4]  type=float

b= [5,6,7,8]  type=float

Peform the following array operations using numpy   
( show both operational use of numpy and functional (example addition operation => + vs addition function => numbpy.add() )

### 3.3.1 addition a+b 

In [4]:
a=np.array([1,2,3,4])
a1=a.astype(float)
b=np.array([5,6,7,8])
b1=b.astype(float)
print a1+b1
print np.add(a1,b1)

[  6.   8.  10.  12.]
[  6.   8.  10.  12.]


### 3.3.2 subtraction a-b

In [5]:
print a1-b1
print np.subtract(a1,b1)

[-4. -4. -4. -4.]
[-4. -4. -4. -4.]


### 3.3.3 multiplication a*b

In [6]:
print a1*b1
print np.multiply(a1,b1)

[  5.  12.  21.  32.]
[  5.  12.  21.  32.]


### 3.3.4 divsion a/b

In [7]:
print a1/b1
print np.divide(a1,b1)

[ 0.2         0.33333333  0.42857143  0.5       ]
[ 0.2         0.33333333  0.42857143  0.5       ]


### 3.3.5 modulo  a%b

In [8]:
print a1%b1
print np.mod(a1,b1)

[ 1.  2.  3.  4.]
[ 1.  2.  3.  4.]


### 3.3.6 power  a^b

In [9]:
print a1**b1
print np.power(a1,b1)

[  1.00000000e+00   6.40000000e+01   2.18700000e+03   6.55360000e+04]
[  1.00000000e+00   6.40000000e+01   2.18700000e+03   6.55360000e+04]


## Homework 3.4
Find your own data and load it into a dataframe

I got my data from Kaggle. 

https://www.kaggle.com/saurograndi/airplane-crashes-since-1908/data 

It contains data of airplane crashes and fatalities since 1908. It would be interesting to see how the number of crashes has varied over time with increase in technology and the number of flights increasing each year.

In [64]:
#Airplane crash data 
apc = pd.read_csv('data/Airplane_Crashes.csv')
apc.head()

Date   Time                            Location  \
0  09/17/1908  17:18                 Fort Myer, Virginia   
1  07/12/1912  06:30             AtlantiCity, New Jersey   
2  08/06/1913    NaN  Victoria, British Columbia, Canada   
3  09/09/1913  18:30                  Over the North Sea   
4  10/17/1913  10:30          Near Johannisthal, Germany   

                 Operator Flight #          Route                    Type  \
0    Military - U.S. Army      NaN  Demonstration        Wright Flyer III   
1    Military - U.S. Navy      NaN    Test flight               Dirigible   
2                 Private        -            NaN        Curtiss seaplane   
3  Military - German Navy      NaN            NaN  Zeppelin L-1 (airship)   
4  Military - German Navy      NaN            NaN  Zeppelin L-2 (airship)   

  Registration cn/In  Aboard  Fatalities  Ground  \
0          NaN     1     2.0         1.0     0.0   
1          NaN   NaN     5.0         5.0     0.0   
2          NaN   NaN     1.0         1.0     0.0   
3          NaN   NaN    20.0        14.0     0.0   
4          NaN   NaN    30.0        30.0     0.0   

                                             Summary  
0  During a demonstration flight, a U.S. Army fly...  
1  First U.S. dirigible Akron exploded just offsh...  
2  The first fatal airplane accident in Canada oc...  
3  The airship flew into a thunderstorm and encou...  
4  Hydrogen gas which was being vented was sucked...

## Homework 3.5
Provide an interesting analysis of the data columns ( frequency or averages )

In [65]:
apc.shape

(5268, 13)

There are 5268 rows and 13 columns

Lets add two columns that are substrings of the date. One will be year and the second will be the decade.

In [90]:
apc["Year"]=apc.Date.str[6:10]
apc["Decade"]=apc.Date.str[6:9]+"0"

In [91]:
apc.head()

Date   Time                            Location  \
0  09/17/1908  17:18                 Fort Myer, Virginia   
1  07/12/1912  06:30             AtlantiCity, New Jersey   
2  08/06/1913    NaN  Victoria, British Columbia, Canada   
3  09/09/1913  18:30                  Over the North Sea   
4  10/17/1913  10:30          Near Johannisthal, Germany   

                 Operator Flight #          Route                    Type  \
0    Military - U.S. Army      NaN  Demonstration        Wright Flyer III   
1    Military - U.S. Navy      NaN    Test flight               Dirigible   
2                 Private        -            NaN        Curtiss seaplane   
3  Military - German Navy      NaN            NaN  Zeppelin L-1 (airship)   
4  Military - German Navy      NaN            NaN  Zeppelin L-2 (airship)   

  Registration cn/In  Aboard  Fatalities  Ground  \
0          NaN     1     2.0         1.0     0.0   
1          NaN   NaN     5.0         5.0     0.0   
2          NaN   NaN     1.0         1.0     0.0   
3          NaN   NaN    20.0        14.0     0.0   
4          NaN   NaN    30.0        30.0     0.0   

                                             Summary  Year Decade  
0  During a demonstration flight, a U.S. Army fly...  1908   1900  
1  First U.S. dirigible Akron exploded just offsh...  1912   1910  
2  The first fatal airplane accident in Canada oc...  1913   1910  
3  The airship flew into a thunderstorm and encou...  1913   1910  
4  Hydrogen gas which was being vented was sucked...  1913   1910

Now that we have decades, lets count the number of crashes for each decade.

In [93]:
pd.value_counts(apc.Decade)

1970    837
1990    775
1960    721
1980    717
1950    596
2000    583
1940    510
1930    323
1920    178
1910     27
1900      1
Name: Decade, dtype: int64

We can see that the largest number of crashes occured in the 1970s. Something interesting is that 1990 is the second highest. Maybe this is becasue there was a larger number of flights in that decade compared to those decades before. Something worth noting is that the number of crashes significantly decreases in the 2000's, we can infer this happended because of increase in technology and safty measurements.